shark_turbine/kernel/wave/expansion.py

```diff
@@ -0,0 +1,324 @@
+    dim_scaling: dict[IndexSymbol, int] = {}
+    hardware_constraints: list[HardwareConstraint] = [
+        constraint
+        for constraint in constraints
+        if isinstance(constraint, HardwareConstraint)
+    ]
+    if len(hardware_constraints) != 1:
+        raise ValueError("Exactly one hardware constraint must be provided")
+
+    idxc = IndexingContext.current()
+    for constraint in constraints:
+        if isinstance(constraint, WorkgroupConstraint):
+            tile_size = idxc.get_static_value(constraint.tile_size)
+            wave_count = hardware_constraints[0].waves_per_block[
+                constraint.workgroup_dim
+            ]
+            mma_size = hardware_constraints[0].mma_matrix_shapes[
+                constraint.workgroup_dim
+            ]
+            if tile_size is None or wave_count is None or mma_size is None:
+                raise ValueError(
+                    "Tile size, wave count and mma size must be statically known"
+                )
+            dim_scaling[constraint.dim] = tile_size // wave_count // mma_size
+    return dim_scaling
+
+
+def _handle_reduction_dim(
+    reduction: Reduction,
+    output: Output,
+    trace: CapturedTrace,
+    dim_scaling: dict[IndexSymbol, int],
+    context: dict[tuple[CustomOp, Symbol, int], CustomOp],
+):
+    # Rediscover iter args
+    # TODO: Register iter args with the reduction initially so accessing them is easier
+    iter_args: list[CustomOp] = []
+    reduction_subgraph = trace.get_subgraph(reduction.subgraph_name)
+    for node in (get_custom(fx_node) for fx_node in reduction_subgraph.nodes):
+        if isinstance(node, IterArg):
+            iter_args.append(node)
+
+    new_outputs = [iter_arg.fx_node for iter_arg in iter_args]
+    # Users of the loop carried nodes will be duplicated
+    for idx, carried_node in enumerate(iter_args):
+        # The initial nodes are expanded in the first dimension, so we start from 1
+        for scale_idx in range(1, dim_scaling[reduction.axis]):
+            for user in carried_node.users:
+                if isinstance(user, Output):
+                    continue
+
+                dims = user.fx_node.expanded_dims
+                dims[reduction.axis] = scale_idx
+                # Temporarily replace the carried arg here to avoid duplicated expansion.
+                # Otherwise we have the following situation:
+                # Suppose we have:
+                #   mma_0_0_0(..., acc_0_0_0)
+                #   mma_0_0_1(..., mma_0_0_0)
+                # Expanding mma_0_0_1 to mma_0_0_2 will trigger expansion of its arg
+                # mma_0_0_0 in dims 0_0_2 as well, effectively duplicating the new node.
+                # To avoid this we temporarily replace the use of it with the original iter_arg
+                # Another option would be to pass a mask of which args to expand, but that
+                # complicates the expansion base case
```

Comment:
Let's chat about this more, I am not sure I fully understand why this is necessary.